logbook-core/src/main/java/org/zalando/logbook/ExtendedLogFormatSink.java

```diff
@@ -0,0 +1,245 @@
+        fieldValMap.put(Field.TIME.value, time);
+        fieldValMap.put(Field.TIME_TAKEN.value, timeTaken);
+        fieldValMap.put(Field.CS_PROTOCOL.value, csProtocol);
+        fieldValMap.put(Field.SC_BYTES.value, scBytes);
+        fieldValMap.put(Field.CS_BYTES.value, csBytes);
+        fieldValMap.put(Field.CLIENT_IP.value, clientIp);
+        fieldValMap.put(Field.SERVER_DNS.value, serverDns);
+        fieldValMap.put(Field.RESP_STATUS.value, status);
+        fieldValMap.put(Field.RESP_COMMENT.value, comment);
+        fieldValMap.put(Field.REQ_METHOD.value, method);
+        fieldValMap.put(Field.REQ_URI.value, uri);
+        fieldValMap.put(Field.REQ_URI_STEM.value, uriStem);
+        fieldValMap.put(Field.REQ_URI_QUERY.value, uriQuery);
+
+        final List<String> outputFields = new ArrayList<>(fields);
+        final String output = outputFields.stream().map(outputField -> {
+                    if (supportedFields.contains(outputField)) {
+                        return fieldValMap.get(outputField);
+                    }
+                    Matcher csHeaderMatcher = CS_HEADER_REGEX.matcher(outputField);
+                    if (csHeaderMatcher.find()) {
+                        String headerKey = csHeaderMatcher.group(1);
+                        return getCsHeader(request, headerKey);
+                    }
+                    Matcher scHeaderMatcher = SC_HEADER_REGEX.matcher(outputField);
+                    if (scHeaderMatcher.find()) {
+                        String headerKey = scHeaderMatcher.group(1);
+                        return getScHeader(response, headerKey);
+                    }
+                    return OMITTED_FIELD;
+                })
+                .reduce((f1, f2) -> String.join(DELIMITER, f1, f2))
+                .orElse("");
+
+        writer.write(correlation, output);
+    }
+
+    private String getCsHeader(HttpRequest request, String key) {
+        return buildHeaderString(request.getHeaders(), key);
+    }
+
+    private String getScHeader(HttpResponse response, String key) {
+        return buildHeaderString(response.getHeaders(), key);
+    }
+
+    private String buildHeaderString(HttpHeaders httpHeaders, String key) {
+        return Optional.of(httpHeaders)
+                .map(headers -> httpHeaders.get(key))
+                .map(values ->
+                        values.stream().reduce(
+                                        (v1, v2) ->
+                                                String.join(HEADER_DELIMITER, v1, v2))
+                                .map(valueStr -> "\"".concat(valueStr).concat("\""))
+                                .orElse(OMITTED_FIELD))
+                .orElse(OMITTED_FIELD);
+    }
+
+    /**
+     * Common supported fields
+     * @see <a href="https://docs.oracle.com/cd/A97329_03/bi.902/a90500/admin-05.htm#634823">Oracle analytical tool log
+     * parsing description</a>
+     */
+    private enum Field {
```

Comment:
It would be nicer if this enum not only holds the field name, but also the logic (as a lambda) how to extract the information. That way the field name and its implementation would be very close to each other. It would also remove the need for the lcoal variables with the comments above.

Reply:
Nice suggestion. Since the logic of getting field values requires data from several places, such as request and response, I have to encapsulate those objects into `FieldParameter` for the lambda usage, please see if that fits. https://github.com/zalando/logbook/pull/1304/commits/45435897f614a8d324e2bf0b1370fb79023f0571